### PR TITLE
GOBBLIN-757: Adding utility functions to support decoration of Avro G…

### PR DESCRIPTION
**Pull request by shirshanka**
…eneric Records

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-757

### Description
- [X] Here are some details about my PR, including screenshots (if applicable):
  Adds two methods to AvroUtils that work on schemas and records. 

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
 additional tests in AvroUtilsTest

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-utility/src/test/java/org/apache/gobblin/util/AvroUtilsTest.java [hunk 1]
```diff
@@ -310,4 +320,157 @@ public void testUnionWithNull() {
...
+            .endRecord();
+    Schema.Field field = new Schema.Field("innerRecord", fieldSchema, "",null);
+    Schema outputRecord = AvroUtils.decorateRecordSchema(inputRecord, Collections.singletonList(field));
+    Assert.assertNotNull(outputRecord.getField("innerRecord"));
+    Assert.assertEquals(outputRecord.getField("innerRecord").schema(), fieldSchema);
+  }
+
+
+  @Test
+  public void testDecorateRecordWithPrimitiveField() {
```
Comment (sv2000) [inline gobblin-utility/src/test/java/org/apache/gobblin/util/AvroUtilsTest.java]: Do you need this test, given the testDecorateRecordWithNestedField?

File: gobblin-utility/src/test/java/org/apache/gobblin/util/AvroUtilsTest.java [hunk 2]
```diff
@@ -310,4 +320,157 @@ public void testUnionWithNull() {
     String stringValue = AvroUtils.getFieldValue(record, "union.string").get().toString();
     Assert.assertEquals(stringValue, "testString");
   }
+
+
+  @Test
+  public void testDecorateSchemaWithSingleField() {
```
Comment (sv2000) [inline gobblin-utility/src/test/java/org/apache/gobblin/util/AvroUtilsTest.java]: Is this test needed given the other testDecorateSchema tests you have below?

File: gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java [hunk 3]
```diff
@@ -834,4 +841,44 @@ public static GenericRecord slowDeserializeGenericRecord(byte[] serializedRecord
...
+    outputSchema.setFields(newOutputFields);
+    copyProperties(inputSchema, outputSchema);
+    return outputSchema;
+  }
+
+  /**
+   * Decorate a {@link GenericRecord} with additional fields and make it conform to an extended Schema
+   * It is the caller's responsibility to ensure that the outputSchema is the merge of the inputRecord's schema
+   * and the additional fields. Since this method is expected to be called in the critical path of processing a record,
+   * this method does not check for safety.
```
[the remark this answers is not in the document]
Comment (shirshanka) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java]: It doesn't check that the outputSchema is indeed the merge of the inputSchema and the additional fields.

File: gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java [hunk 4]
```diff
@@ -74,6 +75,7 @@
 import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.google.common.io.Closer;
+import com.sun.istack.Nullable;
```
Comment (autumnust) [inline gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java]: Is this the expected imports?

File: gobblin-utility/src/test/java/org/apache/gobblin/util/AvroUtilsTest.java (hunk 2 above)
Comment (shirshanka) [reply gobblin-utility/src/test/java/org/apache/gobblin/util/AvroUtilsTest.java]: Wanted to ensure that primitive type works independently of the nested record type. Didn't want to assume that Avro treats both the same way.

File: gobblin-utility/src/test/java/org/apache/gobblin/util/AvroUtilsTest.java (hunk 1 above)
Comment (shirshanka) [reply gobblin-utility/src/test/java/org/apache/gobblin/util/AvroUtilsTest.java]: Same reason as above for schemas.

File: gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java (hunk 4 above)
Comment (shirshanka) [reply gobblin-utility/src/main/java/org/apache/gobblin/util/AvroUtils.java]: thanks for catching. fixed.

**Pull request closed without merging**